### DISCUSSION TITLE
Setting MQTT_TLS_FINGERPRINT to TRUE in user_config_override.h does not enable "force TLS fingerprint validation"

### DISCUSSION THREAD

**Discussion by therobveiller** [arendst/Tasmota#13992] [Ideas]
I am new to GitHub and have no idea how to make a branch and fill in the one-line code addition below, so I am adding it here so that somebody who know how to can. The "issues" template for tasmota doesn't seem to have the relevant sections requests for this sort of code issue as far as I can see, only bugs that happen under certain condition, so I am reporting it here - please don't shoot me!

I use the tls fingerprint method in my custom builds for MQTT TLS authentication. Since the addition of both methods in the same code introduced in V10, the new command of _setoption132 1_ has been added to force the fingerprint method via the console (and this works, but...).
I set many parameters for my custom build via _user_config_override.h_ as I'm supposed to do. This is so that I don't have to issue a vast number of console commands to set up a new device from a standard issue build, as well as adding and removing I2C devices etc to keep the code small but customised to my own needs.

The issue I am having is that the new setting of _MQTT_TLS_FINGERPRINT_ (set to _false_ by default), should if set to TRUE in my _config_override_, be the equivalent of issuing a _setoption132 1_ for new builds and/or if I issue a settings reset to a given device, but it isn't. i.e. if my override code sets MQTT_TLS_ENABLED to TRUE and I flash a new device or issue a reset all settings, then setoption132  still has the value of 0 instead of 1, consequently tls fails to authenticate as it isn't being force to use fingerprint as it should be without me issuing a setoption command.

Looking at the code of settings.ino I can see corresponding flag defintions for the relevant _config_override_ friendly names (e.g. _flag4.mqtt_tls |= MQTT_TLS_ENABLED;_). I do NOT see any such reference for the variable _MQTT_TLS_FINGERPRINT_ in _settings.ino_, (or indeed anywhere else). I would expect to find a line saying something like this
_flag5.tls_use_fingerprint |= MQTT_TLS_FINGERPRINT;_  but there isn't one.
If I add the line of code above, the the builds then work as I would expect, namely setting MQTT_TLS_FINGERPRINT to TRUE in my config_override gives the required behaviour of forcing _setoption132_ to a 1 value when a settings reset is issued on a device or a new device is set up for  the very first time.

Rob

**Comment by s-hadinger** [comment]
Right. I will add it

**Reply by therobveiller** [reply]
Thank-you

**Comment by s-hadinger** [comment]
This is now merged. Thanks for reporting this issue.

**Comment by rapi3** [comment]
Not sure if it is related but compiled 10.1.0.1 and it fail to connect by TLS with error 296
Downgraded to previous 10.0.0.4 I have and TLS worked.

**Reply by s-hadinger** [reply]
Can you share the endpoint ? In PM via Discord if you prefer

**Reply by rapi3** [reply]
I don't have/use Discord... only telegram.
It is my rpi server and firewall it is set to accept connection on that port only from few IPs=DNS so no public access.
mosquitto version 1.3.4 (build date 2019-10-26 22:21:58+0000)

![Screenshot_2021-12-11_18-46-48s](https://user-images.githubusercontent.com/18531150/145686601-e046c9aa-0d0c-4cc8-9c22-deea262f69d0.png)

**Reply by s-hadinger** [reply]
Are you doing fingerprint validation?

**Reply by rapi3** [reply]
nope I used also in console MqttFingerprint " to be sure but no luck after restart.
compiled with this settings for TLS:
#define USE_MQTT_TLS
#undef  MQTT_HOST
#define MQTT_HOST  "yyyyyyyyy"
#undef  MQTT_PORT
#define MQTT_PORT  8883
#undef  MQTT_TLS_ENABLED
#define MQTT_TLS_ENABLED  true
#undef  MQTT_USER
#define MQTT_USER  "xxxx"
#undef  MQTT_PASS
#define MQTT_PASS  "zzzzz"

**Reply by s-hadinger** [reply]
What is the output for `so132` and `mqttfingerprint1` ?

**Reply by rapi3** [reply]
so132 OFF
MqttFingerprint1": "44 7A 46 77 A0 AE 65 5A CC 12 EB CA 98 D0 43 B9 C5 42 32 14

p.s.
tested with so132 1 and 0 and no luck
erased mqttfingerprint and because can't establish TLS connection it remain 00 00 00

**Reply by rapi3** [reply]
I did some test on another newer version of mosquitto and it is connecting; strange.

Latest TLS not work on this is mosquitto server on rpi:
● mosquitto.service - LSB: mosquitto MQTT v3.1 message broker
mosquitto version 1.3.4

Latest TLS work on this is mosquitto server on ubuntu:
● mosquitto.service - Mosquitto MQTT v3.1/v3.1.1 Broker
mosquitto version 1.6.9

**Reply by s-hadinger** [reply]
Oh. I removed the legacy non-elliptic ciphers. Maybe the old version did not support the newer EC ciphers. 

The previous RSA ciphers were completely outdated

**Reply by rapi3** [reply]
unfortunately I can't update rpi OS and mqtt server remote for next 6 month as I am abroad. :(
ps:
this is the error in log: OpenSSL Error: error:1408A0C1:SSL routines:SSL3_GET_CLIENT_HELLO:no shared cipher

**Comment by therobveiller** [comment]
My user_config_override.h section for TLS MQTT with fingerprint validation (as a working example) - thought it might be useful for anyone reading this thread. Fingerprint of my local MQTT server obtained as per Tasmota documentation and then pasted into my code.

I am using Mosquitto 2.0.14 for Windows, running as a background service on WIndows 10 Pro release 21H1


// -- MQTT ----------------------------------------
#ifdef MQTT_HOST
  #undef MQTT_HOST
#endif
#define MQTT_HOST              "aaa.bbb.ccc.ddd"     // [MqttHost]

#ifdef MQTT_PORT
  #undef MQTT_PORT
#endif
#define MQTT_PORT              8883              // [MqttPort] MQTT port

#ifdef MQTT_USER
  #undef MQTT_USER
#endif
#define MQTT_USER              "my_MQTT_username"      // [MqttUser] MQTT user

#ifdef MQTT_PASS
  #undef MQTT_PASS
#endif
#define MQTT_PASS              "my_MQTT_password"       // [MqttPassword] MQTT password

// -- MQTT - Options ------------------------------
#ifndef USE_MQTT_TLS
  #define USE_MQTT_TLS                             // Use TLS for MQTT connection (+34.5k code, +7.0k mem and +4.8k additional during connection handshake)
#endif

#ifdef MQTT_TLS_ENABLED
  #undef MQTT_TLS_ENABLED
#endif
#define MQTT_TLS_ENABLED       true             // [SetOption103] Enable TLS mode (requires TLS version)

#ifdef MQTT_TLS_FINGERPRINT
  #undef MQTT_TLS_FINGERPRINT
#endif
#define MQTT_TLS_FINGERPRINT   true              // [SetOption132] Force TLS fingerprint validation instead of CA (requires TLS version)

#ifdef MQTT_FINGERPRINT1
  #undef MQTT_FINGERPRINT1
#endif
#define MQTT_FINGERPRINT1      0xF4,0xF4,0xF4,0xF4,0xF4,0xF4,0F4,0xF4,0xF4,0xF4,0xF4,0xF4,0xF4,0xF4,0xF4,0xF4,0xF4,0xF4,0xF4,0xF4  // [MqttFingerprint1] (not my actual fingerprint obviously)

**Comment by boozeman** [comment]
Hi,

Just tested with 10.1.0.1 own TLS-build and it is working ok.

Home assistant running on Debian 11 with stand-alone Mosquitto 2.0.11-1 With self-signed CA setup by these instructions [Self signed Mosquitto](https://tasmota.github.io/docs/Self-signed-Mosquitto/)

- I made certificates for the device (Generic 4x relayboard with ESP8266EX chip)
- Configured MQTT to use TLS
- Got SSL error 56 as expected
- set TLSKey1 key for cert
- set TLSKey2  key for cert
- setoption132 1
- Fingerprint autolearned, so no need for mqttfingerprint1 00

user_config_override.h is somewhat minimal.
```
// TLS Stuff here 
#ifndef USE_MQTT_TLS
#define USE_MQTT_TLS
//  #define USE_MQTT_TLS_CA_CERT                    // Force full CA validation instead of fingerprints, slower, but simpler to use.  (+2.2k code, +1.9k mem during connection handshake)
#define USE_MQTT_AWS_IOT                            // This includes the LetsEncrypt CA in tasmota_ca.ino for verifying server certificates
#define USE_MQTT_TLS_FORCE_EC_CIPHER                // Force Elliptic Curve cipher (higher security) required by some servers (automatically enabled with USE_MQTT_AWS_IOT) (+11.4k code, +0.4k mem)
#endif
```
